File: uniquePathsIII.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */
#include"cpptools.h"
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>
#include<queue>

using namespace std;

//#define testMod

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod
class Solution {
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        for(int i=0;i<(int)grid.size();i++)
        {
            for(int j=0;j<(int)grid[0].size();j++)
                if(grid[i][j]==1)
                    return findPath(i,j,grid);
        }
        return -1; // will not go to here
    }

    int findPath(int x,int y,vector<vector<int>>& grid)
    {
        if(x<0||x>=(int)grid.size()||y<0||y>=(int)grid[0].size())
            return 0;
        if(grid[x][y]==-1)
            return 0;
        if(grid[x][y]==2)
        {
            if(check(grid))
                return 1;
            return 0;
        }
        grid[x][y]=-1;
        int out=findPath(x+1,y,grid)+findPath(x-1,y,grid)+
            findPath(x,y+1,grid)+findPath(x,y-1,grid);
        grid[x][y]=0;
        return out;
    }
    
    bool check(vector<vector<int>>& grid)
    {
        for(int i=0;i<(int)grid.size();i++)
        {
            for(int j=0;j<(int)grid[0].size();j++)
            {
                if(grid[i][j]==0)
                    return false;
            }
        }
        return true;
    }
            
private:
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    vector<vector<int>> grid={{1,0,0,0},{0,0,0,0},{0,0,2,-1}};
    cout<<sl.uniquePathsIII(grid)<<endl;
#endif 
    return 0;
}

File: uniquePathsIII.py
from typing import List


class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        c0 = 0
        aa = len(grid)
        bb = len(grid[0])

        for i in range(aa):
            for j in range(bb):
                if grid[i][j] == 0:
                    c0 += 1
                elif grid[i][j] == 1:
                    sa = i
                    sb = j

        def dfs(a, b, k):
            if a < 0 or a >= aa or b < 0 or b >= bb or grid[a][b] == -1:
                return 0
            res = 0

            if grid[a][b] == 2:
                if k == 0:
                    return 1

                return 0
            v = grid[a][b]
            grid[a][b] = -1
            k -= 1
            res += dfs(a + 1, b, k)
            res += dfs(a - 1, b, k)
            res += dfs(a, b + 1, k)
            res += dfs(a, b - 1, k)
            grid[a][b] = v

            return res

        return dfs(sa, sb, c0 + 1)


grid = [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 2, -1]]
grid = [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 2]]
grid = [[0, 1], [2, 0]]
print(Solution().uniquePathsIII(grid))
